### PR TITLE
Add ZStream right, some & orElse variants

### PR DESCRIPTION
**Pull request by luis3m**
Add useful transformations for error & successful channel just like ZIO.

### CODE REVIEW

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 1]
```diff
@@ -2133,6 +2167,21 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
...
+    self.mapError(Some(_)).rightOrFail(None)
+
+  /**
+   * Fails with fiven error 'e' if value is `Left`.
+   */
+  final def rightOrFail[O1, O2, E1 >: E](e: => E1)(implicit ev: O <:< Either[O1, O2]): ZStream[R, E1, O2] =
+    self.flatMap(ev(_) match {
+      case Right(value) => ZStream.succeed(value)
+      case Left(_)      => ZStream.fail(e)
+    })
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala:2183]: Would it be possible to use `mapM` here? It'd be vastly more efficient as:
```suggestion
    self.mapM(ev(_) match {
      case Right(value) => ZIO.succeed(value)
      case Left(_)      => ZIO.fail(e)
    })
```

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 2]
```diff
@@ -2325,6 +2374,27 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
...
+    map(_.getOrElse(default))
+
+  /**
+   * Extracts the optional value, or fails with the given error 'e'.
+   */
+  final def someOrFail[O2, E1 >: E](e: => E1)(implicit ev: O <:< Option[O2]): ZStream[R, E1, O2] =
+    self.flatMap(ev(_) match {
+      case Some(value) => ZStream.succeed(value)
+      case None        => ZStream.fail(e)
+    })
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala:2396]: Similarly to `rightOrFail` - could we use mapM here?

File: streams/shared/src/main/scala/zio/stream/ZStream.scala (hunk 1 above)
Comment (luis3m) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala:2183]: Let me take a look at that

**Comment by luis3m** [comment]
@iravid as suggested, replaced `flatMap` with `mapM`

**Pull request merged**